consensus/parlia/parlia.go

```diff
@@ -317,6 +330,153 @@ func (p *Parlia) VerifyHeaders(chain consensus.ChainHeaderReader, headers []*typ
+	if len(header.Extra) <= extraVanity+extraSeal {
+		return nil, nil
+	}
+
+	if !chainConfig.IsBoneh(header.Number) {
+		return nil, nil
+	}
+
+	var attestationBytes []byte
+	if header.Number.Uint64()%parliaConfig.Epoch != 0 {
+		attestationBytes = header.Extra[extraVanity : len(header.Extra)-extraSeal]
+	} else {
+		num := int(header.Extra[extraVanity])
+		if len(header.Extra) <= extraVanity+extraSeal+validatorNumberSizeAfterBoneh+num*validatorBytesLengthAfterBoneh {
+			return nil, nil
+		}
+		start := extraVanity + validatorNumberSizeAfterBoneh + num*validatorBytesLengthAfterBoneh
+		end := len(header.Extra) - extraVanity
+		attestationBytes = header.Extra[start:end]
+	}
+
+	var attestation types.VoteAttestation
+	if err := rlp.Decode(bytes.NewReader(attestationBytes), &attestation); err != nil {
+		return nil, fmt.Errorf("block %d has finality info, decode err: %s", header.Number.Uint64(), err)
+	}
+	return &attestation, nil
+}
+
+func getSignRecentlyLimit(blockNumber uint64, validatorsNumber int, chainConfig *params.ChainConfig) int {
+	limit := validatorsNumber/2 + 1
+	if chainConfig.IsBoneh(big.NewInt(1).SetUint64(blockNumber)) {
+		limit = validatorsNumber*2/3 + 1
+	}
+	return limit
+}
+
+func (p *Parlia) verifyVoteAttestation(chain consensus.ChainHeaderReader, header *types.Header, parents []*types.Header) error {
+	attestation, err := getVoteAttestationFromHeader(header, p.chainConfig, p.config)
+	if err != nil {
+		return err
+	}
+	if attestation == nil {
+		return nil
+	}
+
+	// Get parent block
+	number := header.Number.Uint64()
+	var parent *types.Header
+	if len(parents) > 0 {
+		parent = parents[0]
+	} else {
+		parent = chain.GetHeader(header.ParentHash, number-1)
+	}
+	if parent == nil || parent.Number.Uint64() != number-1 || parent.Hash() != header.ParentHash {
+		return consensus.ErrUnknownAncestor
+	}
+
+	// The target block should be direct parent
+	targetNumber := attestation.Data.TargetNumber
+	targetHash := attestation.Data.TargetHash
+	if targetNumber != parent.Number.Uint64() || targetHash != parent.Hash() {
+		return fmt.Errorf("invalid attestation, target mismatch, block: %d, target block: %d", number, targetNumber)
+	}
+
+	// The source block should be the highest justified block
+	sourceNumber := attestation.Data.SourceNumber
+	sourceHash := attestation.Data.SourceHash
+	justified := p.GetHighestJustifiedHeader(chain, parent)
+	if justified == nil {
+		return fmt.Errorf("no justified block found")
+	}
+	if sourceNumber != justified.Number.Uint64() || sourceHash != justified.Hash() {
+		return fmt.Errorf("invalid attestation, source mismatch, block: %d, source block: %d", number, sourceNumber)
+	}
+
+	// The snapshot should be the targetNumber-1 block's snapshot.
+	if len(parents) > 2 {
+		parents = parents[2:]
+	} else {
+		parents = nil
+	}
+	snap, err := p.snapshot(chain, parent.Number.Uint64()-1, parent.ParentHash, parents)
```

Comment:
This will cause under-flow, if `parent.Number.Uint64()` is zero, then under-flow happens.

Reply:
parent.Number.Uint64() won't be 0, the first attestation will be earliest included in the 2nd block, the parent block number at least 1, otherwise, the attestation will be nil, and this function will return earlier, have no chance to get here.